cpp/include/cuspatial/experimental/haversine.hpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/constants.hpp>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+namespace cuspatial {
+
+/**
+ * brief Compute haversine distances between points in set A to the corresponding points in set B.
+ *
+ * Computes N haversine distances, where N is `std::distance(a_lon_first, a_lon_last)`. One distance
+ * is computed for each (a_lon[i], a_lat[i]) and (b_lon[i], b_lat[i]) point pair. `distance_first`
+ * must be an iterator to output storage allocated for N distances.
+ *
+ * All iterators must have the same floating-point `value_type`.
+ *
+ * https://en.wikipedia.org/wiki/Haversine_formula
+ *
+ * @param[in]  a_lon_first: beginning of range of longitude of points in set A
+ * @param[in]  a_lon_last: end of the range of longitude of points in set A
+ * @param[in]  b_lon_first: beginning of range of longitude of points in set B
+ * @param[in]  b_lat_first: beginning of range of latitude of points in set B
+ * @param[out] distance_first: beginning of output range of haversine distances
+ * @param[in] radius: radius of the sphere on which the points reside. default: 6371.0
+ *            (approximate radius of Earth in km)
+ *
+ * @tparam LonItA must meet the requirements of [LegacyForwardIterator][LinkLFI] and be
+ * device-accessible.
+ * @tparam LatItA must meet the requirements of [LegacyForwardIterator][LinkLFI] and be
+ * device-accessible.
+ * @tparam LonItB must meet the requirements of [LegacyForwardIterator][LinkLFI] and be
+ * device-accessible.
+ * @tparam LatItB must meet the requirements of [LegacyForwardIterator][LinkLFI] and be
+ * device-accessible.
+ * @tparam OutputIt must meet the requirements of [LegacyForwardIterator][LinkLFI] and be
+ * device-accessible.
```

Comment:
A Forward Iterator only has the `operator++`. These need to be a [`RandomAccessIterator`](https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator) to have an `operator[]`. You can assume that pretty much any parallel algorithm will require a random access iterator because you'll almost never be traversing an iterator serially via `++`.

Reply:
I modeled this after the versions of std::transform that take an exec policy -- they all require LegacyForwardIterator (which I thought was strange, but didn't question). I looked at the Thrust docs but they are out of date and point to old SGI-doc iterators that are no longer accessible.

Reply:
Jake and I discussed offline. This should be LegacyRandomAccessIterator. The reason that std:: algorithms use LegacyForwardIterator (and Thrust uses LegacyInputIterator) is historical, and the parallel algorithms library additions in C++17 made a point of not changing this. But implementations are allowed to fall back to sequential execution or make a copy of input/output ranges. We don't want to specify that fallback is an option, so we think we should prescribe random accessibility. 

Note that this will make the interface not directly usable on things like `std::list` and other non-random-accessible containers.   I don't think this is an issue.

Reply:
For future reference, the new (good) Thrust docs are at https://nvidia.github.io/thrust